include/continuable/detail/other/coroutines.hpp

```diff
@@ -117,16 +121,24 @@ class awaitable {
 
   /// Suspend the current context
   // TODO Convert this to an r-value function once possible
-  void await_suspend(coroutine_handle<> h) {
+  coroutine_handle<> await_suspend(coroutine_handle<> h) {
     assert(result_.is_empty());
     // Forward every result to the current awaitable
     std::move(continuable_)
         .next([h, this](auto&&... args) mutable {
           assert(result_.is_empty());
           result_ = result_t::from(std::forward<decltype(args)>(args)...);
-          h.resume();
+
+          // If nullptr, it means that the promise was not resolved
+          // immediately (i.e., the awaitable await_suspend method
+          // has already returned). That means we must call the
+          // resume function from the continuation chain.
+          if(handle_.exchange(h) == nullptr) {
+            return h.resume();
+          }
         })
         .done();
+    return handle_.exchange(nullptr);
```

Comment:
Is it valid to return an invalid coroutine here?

Reply:
I think the right way to do it would be to let this function return true/false and use an atomic bool instead of std::atomic<coroutine_handle<>>, since you conditionally return the same coroutine that was passed to await_suspend:

https://en.cppreference.com/w/cpp/language/coroutines
> if await_suspend returns bool,
the value true returns control to the caller/resumer of the current coroutine
the value false resumes the current coroutine.

This also solves potential issues when returning an invalid coroutine_handle.

Reply:
I agree with returning a boolean. I wasn't too happy having an atomic coroutine_handle, as I didn't find any guarantee that it would be lock free, but I actually didn't consider return a boolean here.